### PR TITLE
Caches can now also store that fact that a value was missing from the underlying data store

### PR DESCRIPTION
**Pull request by stkem**

### CODE REVIEW

File: kifi-backend/modules/common/app/com/keepit/common/cache/FortyTwoCache.scala [hunk 1]
```diff
@@ -202,16 +225,23 @@ trait FortyTwoCache[K <: Key[T], T] extends ObjectCache[K, T] {
...
+    val valueOpt = try rawValueOpt.asInstanceOf[Option[Option[Any]]] catch {
+      case e: java.lang.ClassCastException => Some(rawValueOpt)
+      case e: Throwable =>
+        repo.onError(HealthcheckError(Some(e), callType = Healthcheck.INTERNAL,
+          errorMessage = Some(s"Failed converting key $key from $repo")))
+        None
+    }
     try {
-      val objOpt = value.map(serializer.reads)
+      val objOpt = valueOpt.map(_.map(serializer.reads))
```
Comment (eishay) [inline kifi-backend/modules/common/app/com/keepit/common/cache/FortyTwoCache.scala:244]: isn't it same as 

``` scala
val objOpt = valueOpt.flatten.map(serializer.reads)
```

?

**Comment by eishay** [comment]
I couldn't find what value is stored in the caches for the None values. Does FortyTwoCachePlugin deal with Option values?

Please add a test for this code. @gmethvin wrote a test tool to check if the test is actually going down to the db layer or stopps at the cache layer.

**Comment by yingjieMiao** [comment]
I guess Eishay meant

sessionProvider.doWithoutCreatingSessions ?

**Comment by eishay** [comment]
yes

**Pull request closed without merging**